src/dhtproto/client/request/internal/Update.d

```diff
@@ -0,0 +1,792 @@
+    import swarm.neo.request.Command;
+    import dhtproto.client.NotifierTypes;
+    import dhtproto.client.internal.NodeHashRanges;
+    import dhtproto.client.internal.SharedResources;
+    import dhtproto.common.Update;
+
+    /// Enum defining three-state return values of some methods.
+    private enum Action
+    {
+        Abort, // An error occurred and the request should be aborted.
+        Retry, // The operation failed, but can be retried (on another node).
+        Success // The operation succeeded.
+    }
+
+    /// Delegate to get access to an event dispatcher for a specific connection.
+    private UseNodeDg use_node;
+
+    /// Delegate to spawn a new request-on-conn. When spawned, Update.handle
+    /// will be called in the new RoC fiber.
+    private NewRequestOnConnDg new_request_on_conn;
+
+    /// Serialised request context.
+    private Update.Context* context;
+
+    /***************************************************************************
+
+        Entry point for the first RoC's logic.
+
+    ***************************************************************************/
+
+    public void handle ( )
+    {
+        auto shared_resources = SharedResources.fromObject(
+            this.context.shared_resources);
+        scope acquired_resources = shared_resources.new RequestResources;
+
+        // Get the list of nodes reponsible for this hash.
+        auto nodes = shared_resources.node_hash_ranges.getNodesForHash(
+            this.context.user_params.args.key,
+            acquired_resources.getNodeHashRangeBuffer());
+
+        // If no node covers the record's hash, exit.
+        if ( nodes.array.length == 0 )
+        {
+            this.context.shared_working.result =
+                Update.SharedWorking.Result.NoNode;
+            return;
+        }
+
+        // Iterate over the list of responsible nodes, in newest to oldest order
+        // of responsibility, trying to get and update the record.
+        bool success, error;
+        foreach ( node_hash_range; nodes.array() )
+        {
+            this.use_node(node_hash_range.addr,
+                ( RequestOnConn.EventDispatcher conn )
+                {
+                    auto ret = this.tryConnection(conn, nodes.array[0].addr,
+                        acquired_resources);
+                    with ( Action ) final switch ( ret )
+                    {
+                        case Success:
+                            this.context.shared_working.result =
+                                Update.SharedWorking.Result.Succeeded;
+                            success = true;
+                            break;
+                        case Abort:
+                            // shared_working.result may either be the default
+                            // value (Error) or may have been set to a specific
+                            // code.
+                            error = true;
+                            break;
```

Comment:
It's weird that this `Abort` actually doesn't abort anything - perhaps we should rename it to something else?